common/transport/amqp/src/amqp_cbs.ts

```diff
@@ -0,0 +1,323 @@
+          forceDetach: () => this._fsm.transition('detached'),
+          putToken: (audience, token, callback) => {
+            this._putTokenQueue.push({
+              audience: audience,
+              token: token,
+              callback: callback
+            });
+          }
+         },
+        attached: {
+          _onEnter: (callback) => {
+            if (callback) {
+              callback();
+            }
+            let tokenOperation = this._putTokenQueue.shift();
+            while (tokenOperation) {
+              this._fsm.handle('putToken', tokenOperation.audience, tokenOperation.token, tokenOperation.callback);
+              tokenOperation = this._putTokenQueue.shift();
+            }
+          },
+          attach: (callback) => callback(),
+          detach: (callback, err) => this._fsm.transition('detaching', callback, err),
+          forceDetach: () => {
+            this._receiverLink.forceDetach();
+            this._senderLink.forceDetach();
+            this._fsm.transition('detached');
+          },
+          putToken: (audience, token, putTokenCallback) => {
+          /*SRS_NODE_AMQP_CBS_16_011: [The `putToken` method shall construct an amqp message that contains the following application properties:
+          ```
+          'operation': 'put-token'
+          'type': 'servicebus.windows.net:sastoken'
+          'name': <audience>
+          ```
+          and system properties of
+          ```
+          'to': '$cbs'
+          'messageId': <uuid>
+          'reply_to': 'cbs'
+          ```
+          and a body containing `<sasToken>`.]*/
+            let amqpMessage = new AmqpMessage();
+            amqpMessage.applicationProperties = {
+              operation: 'put-token',
+              type: 'servicebus.windows.net:sastoken',
+              name: audience
+            };
+            amqpMessage.body = token;
+            amqpMessage.properties = {
+              to: '$cbs',
+              messageId: uuid.v4(),
+              reply_to: 'cbs'
+            };
+
+            let outstandingPutToken: PutTokenOperation = {
+              putTokenCallback: putTokenCallback,
+              expirationTime: Math.round(Date.now() / 1000) + this._putToken.numberOfSecondsToTimeout,
+              correlationId: amqpMessage.properties.messageId
+            };
+
+            this._putToken.outstandingPutTokens.push(outstandingPutToken);
+            //
+            // If this is the first put token then start trying to time it out.
+            //
+            if (this._putToken.outstandingPutTokens.length === 1) {
+              this._putToken.timeoutTimer = setTimeout(this._removeExpiredPutTokens.bind(this), this._putToken.putTokenTimeOutExaminationInterval);
+            }
+            /*Codes_SRS_NODE_AMQP_CBS_16_012: [The `putToken` method shall send this message over the `$cbs` sender link.]*/
+            this._senderLink.send(amqpMessage, (err) => {
+              if (err) {
```

Comment:
>if (err) [](start = 14, length = 8)

It feels like we should be doing something here if the send succeeds. Shouldn't we call putTokenCallback at the very least?

Reply:
we call the callback only when we get the putToken response on the CBS receiver link.

---
In reply to: [139546668](https://github.com/Azure/azure-iot-sdk-node/pull/83#discussion_r139546668) [](ancestors = 139546668)